politeiad/politeiad.go

```diff
@@ -1002,12 +1177,83 @@ func _main() error {
 			if _, found := p.plugins[v.ID]; found {
 				return fmt.Errorf("duplicate plugin: %v", v.ID)
 			}
-
 			p.plugins[v.ID] = convertBackendPlugin(v)
+
+			// Register plugin with the cache
+			cp := convertBackendPluginToCache(v)
+			err := p.cache.RegisterPlugin(cp)
+			if err == cache.ErrWrongPluginVersion {
+				p.cfg.BuildCache = true
+			} else if err != nil {
+				return fmt.Errorf("cache register plugin '%v': %v",
+					cp.ID, err)
+			}
+
+			// Setup the cache plugin tables
+			err = p.cache.PluginSetup(cp.ID)
+			if err != nil {
+				return fmt.Errorf("cache plugin setup '%v': %v",
+					cp.ID, err)
+			}
+
 			log.Infof("Registered plugin: %v", v.ID)
 		}
 	}
 
+	// Build the cache
+	if p.cfg.BuildCache {
+		// Fetch all versions of all records from the inventory and
+		// use them to build the cache
+		vetted, unvetted, err := p.backend.Inventory(0, 0, true, true)
+		if err != nil {
+			return fmt.Errorf("backend inventory: %v", err)
+		}
+
+		inv := make([]cache.Record, 0, len(vetted)+len(unvetted))
+		for _, r := range vetted {
+			inv = append(inv, p.convertBackendRecordToCache(r))
+		}
+		for _, r := range unvetted {
+			inv = append(inv, p.convertBackendRecordToCache(r))
+		}
+
+		// Build the records cache
+		err = p.cache.Build(inv)
+		if err != nil {
+			return fmt.Errorf("build cache: %v", err)
+		}
+
+		// Build the cache for plugins
```

Comment:
eew

Reply:
Anything in particular?

Reply:
Just the XXX needs to be resolved. Needs an issue written up.